linera-core/src/client/mod.rs

```diff
@@ -2460,6 +2339,14 @@ where
     ) -> Result<ClientOutcome<Option<ConfirmedBlockCertificate>>, ChainClientError> {
         let identity = self.identity().await?;
         let mut info = self.chain_info_with_manager_values().await?;
+        {
+            let state = self.state();
+            ensure!(
+                state.block_hash() == info.block_hash
+                    && state.next_block_height() == info.next_block_height,
+                ChainClientError::BlockProposalError("The chain is not synchronized.")
+            );
+        }
```

Comment:
I really dislike the fact that we need to check and re-check every moment whether we can do something or not. OK, you've added this now but how can you be sure that in 20 lines down below (example) we shouldn't re-check again?

Reply:
I agree; `ChainClientState` was introduced in https://github.com/linera-io/linera-protocol/pull/2133, where `ChainClient` was made `Clone`. Since then it is now possible that in another task another `ChainClient` is operating on the same chain, which removed an assumption all this code was written under. (We didn't fully realize that back then.)

`ChainClientState` was meant to partly address this, but its lock is currently not being held long anywhere, and it's difficult to lock it for longer without introducing deadlocks.

I believe we need to work towards getting back to a non-clonable `ChainClient` again, but that's beyond the scope of this PR.

Anyway, the fields of `ChainClientState` originally represented the information that had come out of the wallet file when the process started, and that will get written back to the wallet file when the process terminates. As long as we call `prepare_chain` before doing anything, these checks should always pass for now.